wandb/integration/kfp/wandb_logging.py

```diff
@@ -0,0 +1,166 @@
+        wandb.termlog(f"Setting config: {name} to {data}")
+
+    def log_input_artifact(name, data, type, run=None):
+        artifact = wandb.Artifact(name, type=type)
+        artifact.add_file(data)
+        run.use_artifact(artifact)
+        wandb.termlog(f"Using artifact: {name}")
+
+    def log_output_scalar(name, data, run=None):
+        if isinstance_namedtuple(data):
+            for k, v in zip(data._fields, data):
+                run.log({f"{func.__name__}.{k}": v})
+        else:
+            run.log({name: data})
+
+    def log_output_artifact(name, data, type, run=None):
+        artifact = wandb.Artifact(name, type=type)
+        artifact.add_file(data)
+        run.log_artifact(artifact)
+        wandb.termlog(f"Logging artifact: {name}")
+
+    def _log_component_file(func, run=None):
+        name = func.__name__
+        output_component_file = f"{name}.yml"
+        components._python_op.func_to_component_file(func, output_component_file)
+        artifact = wandb.Artifact(name, type="kubeflow_component_file")
+        artifact.add_file(output_component_file)
+        run.log_artifact(artifact)
+        wandb.termlog(f"Logging component file: {output_component_file}")
+
+    # Add `mlpipeline_ui_metadata_path` to signature to show W&B run in "ML Visualizations tab"
+    sig = signature(func)
+    no_default = []
+    has_default = []
+
+    for param in sig.parameters.values():
+        if param.default is param.empty:
+            no_default.append(param)
+        else:
+            has_default.append(param)
+
+    new_params = tuple(
+        (
+            *no_default,
+            Parameter(
+                "mlpipeline_ui_metadata_path",
+                annotation=OutputPath(),
+                kind=Parameter.POSITIONAL_OR_KEYWORD,
+            ),
+            *has_default,
+        )
+    )
+    new_sig = sig.replace(parameters=new_params)
+    new_anns = {param.name: param.annotation for param in new_params}
+
+    def decorator(func):
+        input_scalars = {}
+        input_artifacts = {}
+        output_scalars = {}
+        output_artifacts = {}
+
+        for name, ann in func.__annotations__.items():
+            if name == "return":
+                output_scalars[name] = ann
+            elif isinstance(ann, output_types):
+                output_artifacts[name] = ann
+            elif isinstance(ann, input_types):
+                input_artifacts[name] = ann
+            else:
+                input_scalars[name] = ann
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            bound = new_sig.bind(*args, **kwargs)
+            bound.apply_defaults()
+
+            mlpipeline_ui_metadata_path = bound.arguments["mlpipeline_ui_metadata_path"]
+            del bound.arguments["mlpipeline_ui_metadata_path"]
+
+            with wandb.init(
+                job_type=func.__name__,
+                group="{{workflow.annotations.pipelines.kubeflow.org/run_name}}",
```

Comment:
Do these fancy format strings work in python 3.7?

Reply:
fstrings should work as of py36!

This specific string `"{{workflow.annotations.pipelines.kubeflow.org/run_name}}"` [is a magic way to get the run name in the component](https://stackoverflow.com/questions/63473716/how-to-obtain-the-kubeflow-pipeline-run-name-from-within-a-component)

Reply:
Ahh, I see.  I thought it was some super fancy python thing, but it's actually a feature of kubeflow.  How does kubeflow replace that?  That's wild...

Reply:
Looks like it's using [these variables](https://argoproj.github.io/argo-workflows/variables/)

I think the actual replacement is happening [here](https://github.com/kubeflow/pipelines/blob/74c7773ca40decfd0d4ed40dc93a6af591bbc190/sdk/python/kfp/compiler/_data_passing_rewriter.py#L11)

Reply:
Hmmm, that makes sense for replacing these when they're a part of the job spec.  In our case it's actual python code.  If they simply regexed our python I would imagine corner cases where replacing the regex would generate invalid code.  I imagine they would need to parse the python AST and then iterate through all strings and look for these patterns...